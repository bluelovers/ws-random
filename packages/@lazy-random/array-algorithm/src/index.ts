
export * from './swapAlgorithm';
export * from './array_rebase';
