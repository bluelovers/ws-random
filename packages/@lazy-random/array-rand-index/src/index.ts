import { _MathRandom } from '@lazy-random/original-math-random';

export function arrayRandIndexByLength(len: number, ...argv: any[])
{
	return Math.floor(_MathRandom() * len)
}

export function arrayRandIndex(array: any[], ...argv: any[])
{
	return arrayRandIndexByLength(array.length)
}

export default arrayRandIndex
