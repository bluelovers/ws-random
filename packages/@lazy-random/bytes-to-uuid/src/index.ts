import { BYTE_TO_HEX_TO_LOWER_CASE } from '@lazy-random/shared-lib';

export function _createBytesToUuidFn(bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	return (buf: ArrayLike<number>, offset?: number) =>
	{
		let i = offset || 0;
		// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
		return ([
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]], '-',
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
			bth[buf[i++]], bth[buf[i++]],
		]).join('');
	}
}

/**
 * @see https://github.com/kelektiv/node-uuid/blob/master/lib/bytesToUuid.js
 */
export function bytesToUuid(buf: ArrayLike<number>, offset?: number, bth = BYTE_TO_HEX_TO_LOWER_CASE)
{
	let i = offset || 0;
	// join used to fix memory issue caused by concatenation: https://bugs.chromium.org/p/v8/issues/detail?id=3175#c4
	return ([
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]], '-',
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
		bth[buf[i++]], bth[buf[i++]],
	]).join('');
}

export default bytesToUuid
