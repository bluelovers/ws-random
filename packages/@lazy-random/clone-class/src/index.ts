/**
 * @todo support typescript
 */
export function getClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o;

	if (thisArgv instanceof RNGClass)
	{
		// @ts-ignore
		o = (thisArgv.__proto__.constructor)
	}
	else
	{
		o = RNGClass
	}

	return o
}

/**
 * @todo support typescript
 */
export function cloneClass<T>(RNGClass: any, thisArgv: any, ...argv: any[]): T
{
	let o = getClass(RNGClass, thisArgv, ...argv)

	// @ts-ignore
	return new o(...argv)
}

export default cloneClass
