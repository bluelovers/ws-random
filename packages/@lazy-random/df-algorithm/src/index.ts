
export { dfBates } from './bates'

export { dfBernoulli } from './bernoulli'
export { dfBinomial } from './binomial'
export { dfExponential } from './exponential'
export { dfGeometric } from './geometric'

export { dfIrwinHall } from './irwin-hall'
export { dfLogNormal } from './log-normal'

export { dfNormal } from './normal'
export { dfPareto } from './pareto'
