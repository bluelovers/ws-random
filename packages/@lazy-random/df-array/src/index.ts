
export { dfArrayIndex } from './array-index'
export { dfArrayShuffle } from './array-shuffle'

export { dfArrayUnique } from './array-unique'
export { dfArrayFill } from './array-fill'

