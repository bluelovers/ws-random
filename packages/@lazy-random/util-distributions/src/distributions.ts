import { IRNGLike } from '@lazy-random/rng-abstract';

export function randIndex(random: IRNGLike, len: number)
{
	return Math.floor(random.next() * len)
}

export function randIndexWithRange(random: IRNGLike, start: number, end: number)
{
	return Math.floor(float(random, start, end))
}

export function float(random: IRNGLike, min: number, max: number)
{
	return random.next() * (max - min) + min;
}

export function int(random: IRNGLike, min: number, max: number)
{
	//return Math.floor(float(random, min, max + 1))
	return randIndexWithRange(random, min, max + 1)
}
