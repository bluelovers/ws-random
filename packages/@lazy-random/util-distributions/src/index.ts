import { float, int, randIndex } from './distributions';

export * from './distributions';

const UtilDistributions = {
	randIndex,
	float,
	int,
};

export default UtilDistributions;
