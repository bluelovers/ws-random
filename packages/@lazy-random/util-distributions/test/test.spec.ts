//@noUnusedParameters:false
/// <reference types="jest" />
/// <reference types="node" />
/// <reference types="expect" />

import { basename, extname } from 'path';
import { newRngMathRandom } from '@lazy-random/util-test';
import { int, randIndex, randIndexWithRange } from '../src/index';
import { ITSTypeAndStringLiteral } from 'ts-type/lib/helper/string';

beforeAll(async () =>
{

});

type IResults = Record<ITSTypeAndStringLiteral<number>, boolean>;

describe(basename(__filename, extname(__filename)), () =>
{
	const testLimit = 1000;

	const rnd = newRngMathRandom();

	test.skip(`dummy`, () => {});

	test(`randIndex`, () =>
	{
		let size = 5;
		let results: IResults = {};
		let actual: number;
		for (let i = 0; i < testLimit; i++)
		{
			actual = randIndex(rnd, size);

			results[actual] ??= true;

			if (actual < 0 || actual >= size)
			{
				break;
			}
		}

		expect(actual).toBeGreaterThan(0);
		expect(actual).toBeLessThan(size);
		expect(results).toMatchSnapshot();

	});

	test(`randIndexWithRange`, () =>
	{
		let min = 1;
		let max = 5;
		let results: IResults = {};
		let actual: number;
		for (let i = 0; i < testLimit; i++)
		{
			actual = randIndexWithRange(rnd, min, max);

			results[actual] ??= true;

			if (actual < min || actual >= max)
			{
				break;
			}
		}

		expect(actual).toBeGreaterThan(0);
		expect(actual).toBeLessThan(max);
		expect(results).toMatchSnapshot();

	});

	test(`int`, () =>
	{
		let min = 1;
		let max = 5;
		let results: IResults = {};
		let actual: number;
		for (let i = 0; i < testLimit; i++)
		{
			actual = int(rnd, min, max);

			results[actual] ??= true;

			if (actual < min || actual > max)
			{
				break;
			}
		}

		expect(actual).toBeGreaterThan(0);
		expect(actual).toBeLessThan(max);
		expect(results).toMatchSnapshot();

	});

})
