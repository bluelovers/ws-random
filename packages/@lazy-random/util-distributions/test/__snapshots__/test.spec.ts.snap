// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test.spec int 1`] = `
{
  "1": true,
  "2": true,
  "3": true,
  "4": true,
  "5": true,
}
`;

exports[`test.spec randIndex 1`] = `
{
  "0": true,
  "1": true,
  "2": true,
  "3": true,
  "4": true,
}
`;

exports[`test.spec randIndexWithRange 1`] = `
{
  "1": true,
  "2": true,
  "3": true,
  "4": true,
}
`;
